core/src/Kokkos_MemoryPool.hpp

```diff
@@ -166,7 +166,8 @@ class MemoryPool {
 
     if (!accessible) {
       Kokkos::Impl::DeepCopy<Kokkos::HostSpace, base_memory_space>(
-          sb_state_array, m_sb_state_array, alloc_size);
+          typename Kokkos::HostSpace::execution_space{}, sb_state_array,
```

Comment:
Note that `base_memory_space::executiuon_space` is more likely to be able to be used in the `DeepCopy`. If the execution space can't enqueue the operation we generally fence the given execution space instance and execute it using a different execution space instance.
Also, using default constructed execution space instances might be problematic in the multi-GPU use case.

Reply:
I couldn't build using `base_memory_space::execution_space` with some backends, iirc.

Reply:
All execution spaces are default-constructible. Do you mind trying again and report related problems here?

Reply:
Ok, this is what I get:

```
error: no instance of constructor "Kokkos::Impl::DeepCopy<Kokkos::HostSpace, MemSpace, ExecutionSpace, std::enable_if_t<<expression>, void>>::DeepCopy [with MemSpace=Kokkos::Cuda::memory_space, ExecutionSpace=Kokkos::HostSpace::execution_space]" matches the argument list
            argument types are: (Kokkos::CudaSpace::execution_space, uint32_t *const, uint32_t *const, const size_t)
        Kokkos::Impl::DeepCopy<Kokkos::HostSpace, base_memory_space>(
        ^
```

Reply:
The overload I would expect to be called is https://github.com/kokkos/kokkos/blob/7ff4042fd2dc1d1fe6ae280179fc693ad6dac6ac/core/src/Cuda/Kokkos_CudaSpace.hpp#L482-L489.
I'm surprised to see `ExecutionSpace=Kokkos::HostSpace::execution_space` and `argument types are: (Kokkos::CudaSpace::execution_space, uint32_t *const, uint32_t *const, const size_t)` at the same time in the error message. Try specifying the third template argument explicitly.

Reply:
It works by passing the execution space explicitly.

Reply:
Eventually, I reverted to using `Kokkos::DefaultHostExecutionSpace`.